keep/step/step.py

```diff
@@ -167,28 +172,34 @@ def _run_single(self):
             result = self._run_single_async()
         # else, just run the provider
         else:
-            try:
-                rendered_providers_parameters = {}
-                for parameter in self.provider_parameters:
-                    rendered_providers_parameters[parameter] = self.io_handler.render(
-                        self.provider_parameters[parameter]
+            for curr_retry_count in range(self.__retry_count + 1):
+                try:
+                    rendered_providers_parameters = {}
+                    for parameter in self.provider_parameters:
+                        rendered_providers_parameters[parameter] = self.io_handler.render(
+                            self.provider_parameters[parameter]
+                        )
+
+                    if self.step_type == StepType.STEP:
+                        step_output = self.provider.query(**rendered_value)
+                        self.context_manager.set_step_context(
+                            self.step_id, results=step_output, foreach=self.foreach
+                        )
+                    else:
+                        self.provider.notify(**rendered_value)
+
+                    extra_context = self.provider.expose()
+                    rendered_providers_parameters.update(extra_context)
+                    self.context_manager.set_step_provider_paremeters(
+                        self.step_id, rendered_providers_parameters
                     )
```

Comment:
One of these may fail and we probably don't want to re-run the step in that case, @shahargl wdyt?

Reply:
@talboren I think we should handle these if we consider this as a part of step execution. If you see below yml example, the `on-failure`, should handle each step, and I think whatever is involved in successful execution of the step, it should be handled, wdyt? 
```
# Alert if a result queried from the DB is above a certain thershold.
alert:
  id: raw-sql-query
  description: Monitor that time difference is no more than 1 hour
  steps:
    - name: get-max-datetime
      provider:
        type: mysql
        config: "{{ providers.mysql-prod }}"
        with:
          # Get max(datetime) from the random table
          query: "SELECT MAX(datetime) FROM demo_table LIMIT 1"
      on-failure:
         retry: 
             count: 4
             interval: 15
  actions:
    - name: trigger-ntfy
      condition:
      - name: threshold-condition
        type: threshold
```

Reply:
@devblin I'm not sure I'm 100% convinced since 
```python
extra_context = self.provider.expose()
rendered_providers_parameters.update(extra_context)
self.context_manager.set_step_provider_paremeters(
    self.step_id, rendered_providers_parameters
)
```
is a bit more "infrastructure"ish -- if something in there fails, it should be some kind of "WTFException" or something like that.

Reply:
@shahargl wdyt ^?

Reply:
so fwiw I think that  the "retry" mechanism is for handling where the *provider* fails and not Keep, so for that matter, only the `notify`/`query` should be in the "retry" block - all other parts are kinda deterministic and retry won't help with that. @talboren

Reply:
@shahargl you mentioned that `notify`/`query` method(s) execution should be handled for retries, so the `on-failure` prop should be available for `action` step too? @talboren